rviz_common/src/rviz_default_plugins/point_cloud_common.cpp

```diff
@@ -0,0 +1,1038 @@
+        M_TransformerInfo::iterator it = transformers_.begin();
+        M_TransformerInfo::iterator end = transformers_.end();
+        for (; it != end; ++it) {
+          const std::string & name = it->first;
+          TransformerInfo & info = it->second;
+
+          setPropertiesHidden(info.xyz_props, name != xyz_transformer_property_->getStdString() );
+          setPropertiesHidden(info.color_props,
+            name != color_transformer_property_->getStdString() );
+        }
+      }
+    }
+
+    new_xyz_transformer_ = false;
+    new_color_transformer_ = false;
+  }
+
+  updateStatus();
+}
+
+void PointCloudCommon::setPropertiesHidden(
+  const QList<rviz_common::properties::Property *> & props,
+  bool hide)
+{
+  for (int i = 0; i < props.size(); i++) {
+    props[i]->setHidden(hide);
+  }
+}
+
+void PointCloudCommon::updateTransformers(
+  const sensor_msgs::msg::PointCloud2::ConstSharedPtr & cloud)
+{
+  std::string xyz_name = xyz_transformer_property_->getStdString();
+  std::string color_name = color_transformer_property_->getStdString();
+
+  xyz_transformer_property_->clearOptions();
+  color_transformer_property_->clearOptions();
+
+  // Get the channels that we could potentially render
+  typedef std::set<std::pair<uint8_t, std::string>> S_string;
+  S_string valid_xyz, valid_color;
+  bool cur_xyz_valid = false;
+  bool cur_color_valid = false;
+  bool has_rgb_transformer = false;
+  M_TransformerInfo::iterator trans_it = transformers_.begin();
+  M_TransformerInfo::iterator trans_end = transformers_.end();
+  for (; trans_it != trans_end; ++trans_it) {
+    const std::string & name = trans_it->first;
+    const PointCloudTransformerPtr & trans = trans_it->second.transformer;
+    uint32_t mask = trans->supports(cloud);
+    if (mask & PointCloudTransformer::Support_XYZ) {
+      valid_xyz.insert(std::make_pair(trans->score(cloud), name));
+      if (name == xyz_name) {
+        cur_xyz_valid = true;
+      }
+      xyz_transformer_property_->addOptionStd(name);
+    }
+
+    if (mask & PointCloudTransformer::Support_Color) {
+      valid_color.insert(std::make_pair(trans->score(cloud), name));
+      if (name == color_name) {
+        cur_color_valid = true;
+      }
+      if (name == "RGB8") {
+        has_rgb_transformer = true;
+      }
+      color_transformer_property_->addOptionStd(name);
+    }
+  }
+
+  if (!cur_xyz_valid) {
+    if (!valid_xyz.empty() ) {
+      xyz_transformer_property_->setStringStd(valid_xyz.rbegin()->second);
+    }
+  }
+
+  if (!cur_color_valid) {
+    if (!valid_color.empty() ) {
+      if (has_rgb_transformer) {
+        color_transformer_property_->setStringStd("RGB8");
+      } else {
+        color_transformer_property_->setStringStd(valid_color.rbegin()->second);
+      }
+    }
+  }
+}
+
+void PointCloudCommon::updateStatus()
+{
+  std::stringstream ss;
+  // ss << "Showing [" << total_point_count_ << "] points from [" << clouds_.size() << "] messages";
```

Comment:
This seems wrong, since it will update the state to empty?

Reply:
This was already commented out in the original rviz code. It seems odd, but I don't understand what the correct behaviour should be - I'll add a TODO and we revisit during refactoring.

Reply:
Sounds good, I couldn't tell the difference in the diff (GitHub didn't see them as rename and change).